shelley-ma/shelley-ma-test/src/Test/Cardano/Ledger/ShelleyMA/Serialisation/Generators.hs

```diff
@@ -1,51 +1,334 @@
+    Arbitrary (RewardAcnt (MaryEra c)),
+    Arbitrary (Addr (MaryEra c)),
+    Arbitrary (TxIn (MaryEra c)),
+    Arbitrary (MultiSig (MaryEra c)),
+    Arbitrary (ScriptHash (MaryEra c)),
+    Arbitrary (TxOut (MaryEra c)),
+    Arbitrary (StrictMaybe SlotNo),
+    Arbitrary SlotNo,
+    Arbitrary Coin,
+    Arbitrary ByteString,
+    Arbitrary Network
+  ) =>
+  Arbitrary (MA.STS.UtxoPredicateFailure (MaryEra c))
+  where
+  arbitrary = genericArbitraryU
+
+instance
+  ( Mock c,
+    PraosCrypto c,
+    Arbitrary (STS.PredicateFailure (UTXO (MaryEra c))),
+    Arbitrary (VKey 'Witness c),
+    Arbitrary (KeyHash 'Witness c),
+    Arbitrary (MetaDataHash (MaryEra c)),
+    Arbitrary (ScriptHash (MaryEra c)),
+    Arbitrary (WitHashes (MaryEra c))
+  ) =>
+  Arbitrary (STS.UtxowPredicateFailure (MaryEra c))
+  where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance
+  ( Mock c,
+    PraosCrypto c,
+    Arbitrary (STS.PredicateFailure (DELEGS (MaryEra c))),
+    Arbitrary (STS.PredicateFailure (UTXOW (MaryEra c)))
+  ) =>
+  Arbitrary (STS.LedgerPredicateFailure (MaryEra c))
+  where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance
+  ( Mock c,
+    PraosCrypto c,
+    Arbitrary (STS.PredicateFailure (LEDGER (MaryEra c)))
+  ) =>
+  Arbitrary (STS.LedgersPredicateFailure (MaryEra c))
+  where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+{-------------------------------------------------------------------------------
+  AllegraEra Generators
+-------------------------------------------------------------------------------}
+
+instance
+  ( Mock c,
+    PraosCrypto c,
+    Arbitrary (Abstract.Value (AllegraEra c)),
+    Arbitrary (TxIn (AllegraEra c)),
+    Arbitrary (StrictSeq (TxOut (AllegraEra c))),
+    Arbitrary (StrictSeq (DCert (AllegraEra c))),
+    Arbitrary (Wdrl (AllegraEra c)),
+    Arbitrary (StrictMaybe (Update (AllegraEra c))),
+    Arbitrary (StrictMaybe (MetaDataHash (AllegraEra c))),
+    Arbitrary (StrictMaybe SlotNo)
+  ) =>
+  Arbitrary (MA.TxBody (AllegraEra c))
+  where
+  arbitrary = genTxBody
+
+instance
+  ( Mock c,
+    PraosCrypto c,
+    Arbitrary (StrictMaybe SlotNo),
+    Arbitrary (MultiSig (AllegraEra c))
+  ) =>
+  Arbitrary (Timelock (AllegraEra c))
+  where
   arbitrary = sizedTimelock maxTimelockDepth
+
+instance
+  Arbitrary (StrictMaybe SlotNo) =>
+  Arbitrary ValidityInterval
+  where
+  arbitrary = genericArbitraryU
+  shrink = genericShrink
+
+instance
+  ( Mock c,
+    PraosCrypto c,
+    Arbitrary (STS.PpupPredicateFailure (AllegraEra c)),
+    Arbitrary (RewardAcnt (AllegraEra c)),
+    Arbitrary (Addr (AllegraEra c)),
+    Arbitrary (TxIn (AllegraEra c)),
+    Arbitrary (MultiSig (AllegraEra c)),
+    Arbitrary (ScriptHash (AllegraEra c)),
+    Arbitrary (TxOut (AllegraEra c)),
+    Arbitrary (StrictMaybe SlotNo),
+    Arbitrary SlotNo,
+    Arbitrary Coin,
+    Arbitrary ByteString,
+    Arbitrary Network,
+    Arbitrary DeltaCoin
```

Comment:
Don't these exist already?

Reply:
these do exist in `EraIndep` - but I'm avoiding to reference that here so that we can use EraIndep as an entry point

Reply:
What's your objection to depending on `EraIndep` here? I would understand objecting to the other direction.

Reply:
good point! I'll do that and mop up the constraints